File: Src/Backend/routes/social.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from models import User, Friendship, Message
from extensions import db
from sqlalchemy import or_
from utils import get_current_user

social_bp = Blueprint('social', __name__)


def _get_friendship_status(user_a, user_b):
    friendship = Friendship.query.filter(
        or_(
            (Friendship.user_id == user_a.id) & (Friendship.friend_id == user_b.id),
            (Friendship.user_id == user_b.id) & (Friendship.friend_id == user_a.id)
        )
    ).first()
    return friendship.status if friendship else 'none'


def _serialize_user(user):
    return {
        "id": user.id,
        "username": user.username,
        "public_id": user.public_id,
        "avatar_url": user.avatar_url
    }


@social_bp.route('/users/search', methods=['GET'])
@jwt_required()
def search_users():
    user = get_current_user()
    query = request.args.get('q', '')

    if not query:
        return jsonify([]), 200

    users = User.query.filter(
        or_(
            User.username.ilike(f'%{query}%'),
            User.public_id.ilike(f'%{query}%')
        )
    ).filter(User.id != user.id).limit(10).all()

    result = []
    for u in users:
        status = _get_friendship_status(user, u)
        entry = _serialize_user(u)
        entry["friend_status"] = status
        result.append(entry)

    return jsonify(result), 200


@social_bp.route('/friends/request', methods=['POST'])
@jwt_required()
def send_friend_request():
    user = get_current_user()
    data = request.get_json()
    friend_id = data.get('friend_id')

    if not friend_id:
        return jsonify({"msg": "Friend ID required"}), 400

    if user.id == friend_id:
        return jsonify({"msg": "Cannot add yourself"}), 400

    existing = Friendship.query.filter(
        or_(
            (Friendship.user_id == user.id) & (Friendship.friend_id == friend_id),
            (Friendship.user_id == friend_id) & (Friendship.friend_id == user.id)
        )
    ).first()

    if existing:
        return jsonify({"msg": "Request already exists or already friends"}), 400

    new_request = Friendship(user_id=user.id, friend_id=friend_id, status='pending')
    db.session.add(new_request)
    db.session.commit()

    return jsonify({"msg": "Request sent"}), 201


@social_bp.route('/friends/accept', methods=['POST'])
@jwt_required()
def accept_friend_request():
    user = get_current_user()
    data = request.get_json()
    request_id = data.get('request_id')

    friendship = Friendship.query.get(request_id)
    if not friendship or friendship.friend_id != user.id:
        return jsonify({"msg": "Request not found"}), 404

    friendship.status = 'accepted'
    db.session.commit()

    return jsonify({"msg": "Request accepted"}), 200


@social_bp.route('/friends', methods=['GET'])
@jwt_required()
def get_friends():
    user = get_current_user()

    friendships = Friendship.query.filter(
        ((Friendship.user_id == user.id) | (Friendship.friend_id == user.id)),
        Friendship.status == 'accepted'
    ).all()

    result = []
    for f in friendships:
        friend = f.friend if f.user_id == user.id else f.user
        unread_count = Message.query.filter_by(
            sender_id=friend.id,
            receiver_id=user.id,
            is_read=False
        ).count()

        entry = _serialize_user(friend)
        entry["unread_count"] = unread_count
        result.append(entry)

    return jsonify(result), 200


@social_bp.route('/friends/pending', methods=['GET'])
@jwt_required()
def get_pending_requests():
    user = get_current_user()

    requests = Friendship.query.filter_by(friend_id=user.id, status='pending').all()
    result = []
    for r in requests:
        result.append({
            "request_id": r.id,
            **_serialize_user(r.user)
        })

    return jsonify(result), 200

File: source/backend/routes/social.js
const express = require('express');
const router = express.Router();
const { authenticateToken } = require('../middleware/auth');

async function getFriendshipStatus(prisma, userAId, userBId) {
    const friendship = await prisma.friendship.findFirst({
        where: {
            OR: [
                { user_id: userAId, friend_id: userBId },
                { user_id: userBId, friend_id: userAId }
            ]
        }
    });
    return friendship ? friendship.status : 'none';
}

function serializeUser(user) {
    return {
        id: user.id,
        username: user.username,
        public_id: user.public_id,
        avatar_url: user.avatar_url
    };
}

// GET /users/search
router.get('/users/search', authenticateToken, async (req, res) => {
    const prisma = req.app.get('prisma');
    const userId = parseInt(req.user.sub);
    const query = req.query.q || '';

    if (!query) {
        return res.status(200).json([]);
    }

    try {
        const users = await prisma.user.findMany({
            where: {
                id: { not: userId },
                OR: [
                    { username: { contains: query } },
                    { public_id: { contains: query } }
                ]
            },
            take: 10
        });

        const result = [];
        for (const u of users) {
            const status = await getFriendshipStatus(prisma, userId, u.id);
            const entry = serializeUser(u);
            entry.friend_status = status;
            result.push(entry);
        }

        res.status(200).json(result);
    } catch (err) {
        console.error(err);
        res.status(500).json({ msg: "Server error" });
    }
});

// POST /friends/request
router.post('/friends/request', authenticateToken, async (req, res) => {
    const prisma = req.app.get('prisma');
    const userId = parseInt(req.user.sub);
    const { friend_id } = req.body;

    if (!friend_id) return res.status(400).json({ msg: "Friend ID required" });
    if (userId === parseInt(friend_id)) return res.status(400).json({ msg: "Cannot add yourself" });

    try {
        const existing = await prisma.friendship.findFirst({
            where: {
                OR: [
                    { user_id: userId, friend_id: parseInt(friend_id) },
                    { user_id: parseInt(friend_id), friend_id: userId }
                ]
            }
        });

        if (existing) {
            return res.status(400).json({ msg: "Request already exists or already friends" });
        }

        await prisma.friendship.create({
            data: {
                user_id: userId,
                friend_id: parseInt(friend_id),
                status: 'pending'
            }
        });

        res.status(201).json({ msg: "Request sent" });
    } catch (err) {
        console.error(err);
        res.status(500).json({ msg: "Server error" });
    }
});

// POST /friends/accept
router.post('/friends/accept', authenticateToken, async (req, res) => {
    const prisma = req.app.get('prisma');
    const userId = parseInt(req.user.sub);
    const { request_id } = req.body;

    try {
        const friendship = await prisma.friendship.findUnique({ where: { id: parseInt(request_id) } });
        if (!friendship || friendship.friend_id !== userId) {
            return res.status(404).json({ msg: "Request not found" });
        }

        await prisma.friendship.update({
            where: { id: parseInt(request_id) },
            data: { status: 'accepted' }
        });

        res.status(200).json({ msg: "Request accepted" });
    } catch (err) {
        console.error(err);
        res.status(500).json({ msg: "Server error" });
    }
});

// GET /friends
router.get('/friends', authenticateToken, async (req, res) => {
    const prisma = req.app.get('prisma');
    const userId = parseInt(req.user.sub);

    try {
        const friendships = await prisma.friendship.findMany({
            where: {
                OR: [
                    { user_id: userId },
                    { friend_id: userId }
                ],
                status: 'accepted'
            },
            include: { user: true, friend: true }
        });

        const result = [];
        for (const f of friendships) {
            const friend = f.user_id === userId ? f.friend : f.user;
            const unread_count = await prisma.message.count({
                where: {
                    sender_id: friend.id,
                    receiver_id: userId,
                    is_read: false
                }
            });

            const entry = serializeUser(friend);
            entry.unread_count = unread_count;
            result.push(entry);
        }

        res.status(200).json(result);
    } catch (err) {
        console.error(err);
        res.status(500).json({ msg: "Server error" });
    }
});

// GET /friends/pending
router.get('/friends/pending', authenticateToken, async (req, res) => {
    const prisma = req.app.get('prisma');
    const userId = parseInt(req.user.sub);

    try {
        const requests = await prisma.friendship.findMany({
            where: { friend_id: userId, status: 'pending' },
            include: { user: true }
        });

        const result = requests.map(r => ({
            request_id: r.id,
            ...serializeUser(r.user)
        }));

        res.status(200).json(result);
    } catch (err) {
        console.error(err);
        res.status(500).json({ msg: "Server error" });
    }
});

module.exports = router;
